src/ModelContextProtocol/Protocol/Messages/ProgressToken.cs

```diff
@@ -0,0 +1,102 @@
+﻿using ModelContextProtocol.Utils;
+using System.ComponentModel;
+using System.Globalization;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace ModelContextProtocol.Protocol.Messages;
+
+/// <summary>
+/// Represents a progress token, which can be either a string or an integer.
+/// </summary>
+[JsonConverter(typeof(Converter))]
+public readonly struct ProgressToken : IEquatable<ProgressToken>
+{
+    /// <summary>The id, either a string or a boxed long or null.</summary>
+    private readonly object? _id;
+
+    /// <summary>Initializes a new instance of the <see cref="ProgressToken"/> with a specified value.</summary>
+    /// <param name="value">The required ID value.</param>
+    public ProgressToken(string value)
+    {
+        Throw.IfNull(value);
+        _id = value;
+    }
+
+    /// <summary>Initializes a new instance of the <see cref="ProgressToken"/> with a specified value.</summary>
+    /// <param name="value">The required ID value.</param>
+    public ProgressToken(long value)
+    {
+        // Box the long. Progress tokens are almost always strings in practice, so this should be rare.
+        _id = value;
+    }
+
+    /// <summary>Gets whether the identifier is uninitialized.</summary>
+    public bool IsDefault => _id is null;
+
+    /// <inheritdoc />
+    public override string? ToString() =>
+        _id is string stringValue ? stringValue :
```

Comment:
Related to https://github.com/modelcontextprotocol/csharp-sdk/pull/146/files#r2020810994, should we be inserting quotes for string values to distinguish between quoted and unquoted numbers?